dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc

```diff
@@ -240,18 +226,28 @@ DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
 
   CUDA_CALL(cudaEventSynchronize(ctx.decode_event));
 
-  // nvJPEG2000 decodes into planar layout
-  bool is_conversion_needed = ctx.shape[2] > 1 || ctx.bpp == 16;
+  DALIImageType format = ctx.shape[2] == 1 ? DALI_GRAY : DALI_RGB;
+  bool is_processing_needed =
+    ctx.shape[2] > 1 ||  // nvJPEG2000 decodes into planar layout
+    ctx.pixel_type != opts.dtype ||
+    format != opts.format ||
+    (ctx.bpp != 8 && ctx.bpp != 16);
+
   try {
```

Comment:
I guess the other way could be to return a cudaError in DecodeJpeg2000, but I don't think it changes much.

I believe Decode only throws if the data passed is somehow invalid (which should be rare or even not happen at all), in case of corrupted jpeg2000 it should just return false.